c7n/resources/ebs.py

```diff
@@ -247,6 +247,63 @@ def process_resource_set(self, client, resource_set):
         return results
 
 
+@Snapshot.filter_registry.register('unused')
+class SnapshotUnusedFilter(Filter):
+    """Filters snapshots based on usage
+    true: snapshot is not used by launch-template, launch-config, or ami.
+    false: snapshot is being used by launch-template, launch-config, or ami.
+    :example:
+    .. code-block:: yaml
+            policies:
+              - name: snapshot-unused
+                resource: ebs-snapshot
+                filters:
+                  - type: unused
+                    value: true
+    """
+    schema = type_schema('unused', value={'type': 'boolean'})
+
+    def get_permissions(self):
+        return list(itertools.chain([
+            self.manager.get_resource_manager(m).get_permissions()
+            for m in ('asg', 'launch-config', 'ami')]))
+
+    def _pull_asg_snapshots(self):
+        asgs = self.manager.get_resource_manager('asg').resources()
```

Comment:
instead of implementing all this logic here, let's extend the asg.LaunchInfo class with a get_snapshot_ids since it already has common logic for abtracting over launch template/launch config.

Reply:
let's leave this as is, i took a deeper look at this, and there's a larger refactoring needed to unify the duplication on block device handling, i counted at least 5 different variations.

Reply:
Gotcha! That's with respect to asg snapshot? Pretty sure, snapshots attached to the AMI doesn't allow deletion until forced.

Reply:
its more about the duplication of logic to extract block devices for args inclusive of, snapshots, ebs out of Amis, all having to handle launch template / launch configuration variations.

Reply:
ie. not specific this to pr, just this adds another variation/duplication, but that cleanup/refactoring should be handled in a separate pr

Reply:
Gotcha! Just fixed the lint error! Tox always tends to fail in my PR